src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,137 @@
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument;
+            if (textDocument == null)
+                return null;
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
+                return null;
+
+            var codeMapper = document.GetLanguageService<IMapCodeService>();
+            if (codeMapper == null)
+                return null;
+
+            var focusLocations = await ConvertFocusLocationsToDocumentAndSpansAsync(
+                document,
+                textDocument,
+                codeMapping.FocusLocations,
+                cancellationToken).ConfigureAwait(false);
+
+            var newDocument = await codeMapper.MapCodeAsync(
+                document,
+                codeMapping.Contents.ToImmutableArrayOrEmpty(),
+                focusLocations,
+                formatMappedCode: true,
+                cancellationToken).ConfigureAwait(false);
+
+            if (newDocument is null)
+                return null;
+
+            var textChanges = (await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false)).ToImmutableArray();
+            if (textChanges.IsEmpty)
+                return null;
+
+            var oldText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textEdits = textChanges.Select(change => ProtocolConversions.TextChangeToTextEdit(change, oldText)).ToArray();
+
+            return (newDocument.GetURI(), textEdits);
+        }
+
+        static async Task<ImmutableArray<(Document, TextSpan)>> ConvertFocusLocationsToDocumentAndSpansAsync(
+            Document document, TextDocumentIdentifier textDocumentIdentifier, LSP.Location[][]? focusLocations, CancellationToken cancellationToken)
+        {
+            if (focusLocations is null)
+                return ImmutableArray<(Document, TextSpan)>.Empty;
+
+            using var _ = ArrayBuilder<(Document, TextSpan)>.GetInstance(out var builder);
+            foreach (var locationsOfSamePriority in focusLocations)
+            {
+                foreach (var location in locationsOfSamePriority)
+                {
+                    // Ignore anything not in target document, which current code mapper doesn't handle anyway
+                    if (!location.Uri.Equals(textDocumentIdentifier.Uri))
+                        continue;
```

Comment:
log failures here.  we've had many uri equality issues so you might encounter weird bugs.